### PR TITLE
Delete DIP after processing, refs #13396

### PR DESCRIPTION
**Pull request by djjuhasz**
- Add a `qtPackageExtractorBase` destructor to delete DIP files after
  processing or when extraction fails
- Add `\AccessToMemory\Path` class and tests
- Use `Path::delete()` to recursively delete the DIP directory
- Pass a logger to `qtPackageExtratorBase` to log DIP deletion
- Remove unused `qtPackageExtractorBase::getFilesFromDirectory()`
  method
- Remove unnecessary autoloader from `test/bootstrap/phpunit.php`

### CODE REVIEW

File: test/phpunit/PathTest.php [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        $directory = [
+            'subdir' => [
+                'foo.txt' => 'foo',
+            ],
+            'empty_dir' => [],
+            'bar.txt' => 'bar',
+        ];
+
+        // setup and cache the virtual file system
+        $this->vfs = vfsStream::setup('root', null, $directory);
```
Comment (jraddaoui) [inline test/phpunit/PathTest.php:44]: Cool!

File: lib/AccessToMemory/Path.php [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+    protected function throwExceptionIfNotAFile()
+    {
+        if (!$this->isFile()) {
+            throw new \RuntimeException(
+                sprintf('%s is not a file', $this->path)
+            );
+        }
+    }
+}
```
Comment (jraddaoui) [inline lib/AccessToMemory/Path.php:175]: You went deep in this abstraction to just delete the directory, hopefully we give it more use in the future.

File: plugins/qtSwordPlugin/lib/qtPackageExtractorBase.class.php [hunk 3]
```diff
@@ -55,6 +57,15 @@ public function __construct(array $options = [])
...
         }
+
+        if (isset($options['logger'])) {
+            $this->logger = $options['logger'];
+        }
+    }
+
+    public function __destruct()
+    {
+        $this->cleanUp();
```
Comment (jraddaoui) [inline plugins/qtSwordPlugin/lib/qtPackageExtractorBase.class.php]: Nitpick: I don't see a point (yet) to have another clean-up method that is only called here.

File: test/bootstrap/phpunit.php [hunk 4]
```diff
@@ -14,14 +14,6 @@
...
 $configuration = ProjectConfiguration::hasActive() ? ProjectConfiguration::getActive() : new ProjectConfiguration(realpath($_test_dir.'/..'));
 
-// autoloader
-$autoload = sfSimpleAutoload::getInstance(sfConfig::get('sf_cache_dir').'/project_autoload.cache');
-$autoload->loadConfiguration(sfFinder::type('file')->name('autoload.yml')->in([
-    sfConfig::get('sf_symfony_lib_dir').'/config/config',
-    sfConfig::get('sf_config_dir'),
-]));
-$autoload->register();
-
```
Comment (jraddaoui) [inline test/bootstrap/phpunit.php:24]: Nice! I always wondered why was this needed.

File: test/phpunit/PathTest.php [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        $path = new Path($this->vfs->url().'/bar.txt');
+        $path->delete();
+        $this->assertFalse($this->vfs->hasChild('bar.txt'));
+
+        // Delete a dir recursively
+        $path = new Path($this->vfs->url().'/subdir');
+        $path->delete(true);
+        $this->assertFalse($this->vfs->hasChild('subdir'));
+    }
+}
```
Comment (jraddaoui) [inline test/phpunit/PathTest.php:204]: Neat!

**Review by jraddaoui — APPROVED** [review]
Looks great @djjuhasz! Maybe a bit too much, but hopefully we'll give a good use to that path abstraction in the future.

File: lib/AccessToMemory/Path.php (hunk 2 above)
Comment (djjuhasz) [reply lib/AccessToMemory/Path.php:175]: Yes, I also needed to recursively delete a directory in PR https://github.com/artefactual/atom/pull/1500.  

In PR #1500 I used [sfToolkit::clearDirectory()](https://github.com/artefactual/atom/blob/938ee91eeff42c168aa82c641b7e979785596634/lib/job/arZipFileDownload.class.php#L84) which I copied from the previous code. When I was working on this fix I examined `sfToolkit::clearDirectory()` and realized that it won't delete a directory that contains `.svn` or `CVS` files anywhere in it's subtree. :(   It's pretty unlikely that a DIP or export directory would have a CVS or .svn file, but it seems to me that it's a non-intuitive behaviour that could be a problem in some cases.

I also found in writing the Path library that the PHP functions like `rmdir()` and `unlink()` will fail without any error if e.g. you try and `unlink()` a directory, or vice versa.   In some cases a silent failure is fine, but there are cases where a it causes a bug. I think it's better in general to consistently throw an exception on failure, and give the developer the choice to catch the exception if they don't want the script to halt.

File: plugins/qtSwordPlugin/lib/qtPackageExtractorBase.class.php (hunk 3 above)
Comment (djjuhasz) [reply plugins/qtSwordPlugin/lib/qtPackageExtractorBase.class.php]: Yeah, fair. I'll move the code into the `__destruct()` method.

**Comment by djjuhasz** [comment]
> Looks great @djjuhasz! Maybe a bit too much, but hopefully we'll give a good use to that path abstraction in the future.

Probably too much. But have you seen the test coverage report for `\AccessToMemory\Path`? 100%! 😆
